tests/tests_unit/test_rules/test_models/test_restrictions.py

```diff
@@ -0,0 +1,143 @@
+            (
+                "valueConstraint:has.complex_property-123(hasValue,test)",
+                "has.complex_property-123",
+                "hasValue",
+                ConceptEntity,
+            ),
+        ],
+    )
+    def test_parse_value_constraints(self, data, expected_property, expected_constraint, expected_value_type):
+        defaults = {"prefix": "ex"} if "Vehicle" in data or "Color" in data or "test" in data else {}
+        result = ConceptPropertyValueConstraint._parse(data, defaults)
+
+        assert result["property_"] == expected_property
+        assert result["constraint"] == expected_constraint
+        assert isinstance(result["value"], expected_value_type)
+
+    def test_parse_named_individual_specific(self):
+        data = "valueConstraint:hasOwner(hasValue,ni:John)"
+        result = ConceptPropertyValueConstraint._parse(data, {})
+        assert str(result["value"]) == "ni:John"
+
+    def test_parse_datatype_specific(self):
+        data = "valueConstraint:hasAge(hasValue,25^^integer)"
+        result = ConceptPropertyValueConstraint.load(data)
+        assert result.value.value == 25
+        assert result.value.datatype == XSD.integer
+
+    @pytest.mark.parametrize(
+        "invalid_data,expected_error",
+        [
+            ("valueConstraint:hasAge(hasValue,25^^invalidtype)", "Invalid value format for datatype"),
+            ("invalidConstraint:hasAge(hasValue,25)", "Invalid value constraint format"),
+        ],
+    )
+    def test_parse_invalid_formats(self, invalid_data, expected_error):
+        with pytest.raises(ValidationError, match=expected_error):
+            ConceptPropertyValueConstraint.load(invalid_data)
+
+
+class TestConceptPropertyCardinalityConstraint:
+    @pytest.mark.parametrize(
+        "data,expected_property,expected_constraint,expected_value,expected_on",
+        [
+            ("cardinalityConstraint:hasChild(minCardinality,2)", "hasChild", "minCardinality", 2, None),
+            ("cardinalityConstraint:hasParent(maxCardinality,2)", "hasParent", "maxCardinality", 2, None),
+            ("cardinalityConstraint:hasSpouse(cardinality,1)", "hasSpouse", "cardinality", 1, None),
+            ("cardinalityConstraint:hasProperty(minCardinality,0)", "hasProperty", "minCardinality", 0, None),
+            (
+                "cardinalityConstraint:has.complex_property-123(maxCardinality,5)",
+                "has.complex_property-123",
+                "maxCardinality",
+                5,
+                None,
+            ),
+        ],
+    )
+    def test_parse_basic_cardinality_constraints(
+        self, data, expected_property, expected_constraint, expected_value, expected_on
+    ):
+        result = ConceptPropertyCardinalityConstraint.load(data)
+
+        assert result.property_ == expected_property
+        assert result.constraint == expected_constraint
+        assert result.value == expected_value
+        assert result.on == expected_on
```

Comment:
Suggest to also assert that the serialization works in reverse as well. 
```suggestion
        assert result.on == expected_on
        
        assert str(result) == data
```